Add create retro and edit retro links to /pairs

mbramson opened the pull request:
Added back the create retro and edit retro links to /pairs as they were removed because of a bug when group-based pairing was added.

Previously, the query that retrieved retros simply looked for a retro (and unfortunately multiple) that was associated with the user and the year and week (through a pair). This worked fine when it was only possible for a user to be in one pair per week, but it broke down once a user could be in many pairs in the same week due to being in multiple groups. The end effect was if you were paired in Group A and Group B, and created a retrospective for Group A, the "create retro" link would change to "edit retro" for both Group A and Group B, and would link to the retro that you created for Group A under both links. This now works properly.

This also fixed a subtle bug due to zipping the user's groups with the user's pairs for a given week when a user wasn't in a pair for all of their groups that week. If that pair that they weren't in was not the last group in the list of groups (order was arbitrary), then the pairs could shift to a different group. Pairs are now connected to groups explicitly, so it is possible for the second group out of three to not have a pair for the current user and have it display properly.

Although not a possible case presently, the /pairs route will now properly display multiple pairs for a group if the user is in multiple pairs.

Closes #126

mbramson committed:
Add retro_for_user_and_pair query

mbramson committed:
Fix potential Enum.zip mismatch

In the logic gathering the pairs for groups in the pair controller, the
logic that simply zips the user's groups with pairs has been replaced
with nested tuples that make sure pairs cannot be incorrectly displayed
next to a group that they are not actually associated with.

mbramson committed:
Attach potential retros to pairs on /pairs

Instead of returning {group, pairs}, return {group, [ {pair, retro} ]}
(a list of pair and retro tuples instead of just a list of pairs). The
retro that is returned can either be a retro that the user is that is
associated with that pair or it can be nil (which means that the user
does not have an existing retrospective for that pair)

The retro does not display in any way on the /pairs index yet

mbramson committed:
Display retrospectives on the /pairs index

mbramson committed:
Fix broken tests for retros on /pairs

mbramson committed:
Replace functions with ex_machina in retro tests

mbramson committed:
Add additional tests cases around two groups

mbramson committed:
Add more concise test

Add Pairmotron.UserGroup schema to the join_through on the User :groups
association to prevent timestamp issues with ex machina.

mbramson commented on web/controllers/pair_controller.ex:
```diff
@@ -29,30 +29,37 @@ defmodule Pairmotron.PairController do
 
-  @spec pairs_for_user_groups(Types.user, integer(), 1..53) :: {[Types.pair], [String.t]}
-  defp pairs_for_user_groups(user, year, week) do
+  @spec groups_with_pairs_for_user(Types.user, integer(), 1..53) :: {[{Types.group, [{Types.pair, Types.retro | nil}]}], [String.t]}
```
So the data structure that we're returning got a bit gnarly. Every group has a list of tuples where that tuple is a pair and possibly a retro (if it exists for that pair and the current user). Definitely open to suggestions on how to make this less mind bending to think about.

ericworkman replied:
Perhaps its time for a struct for a `PairSession`, which has users and a retro and group attached to it? That way, you are just passing around a list of PairSessions?

mbramson replied:
That would definitely make this easier to reason about.

mbramson replied:
What do you think about just adding a @type for pair_session that does the same thing? No explicit struct since the PairController is the only place this is used, but still gain a lot of clarity in the type_specs. I'm not sure its worth encoding this structure as a return type, but I'm not sure. Thoughts?

ericworkman replied:
Its also used in the group pair controller iirc. At least a @type would be a good start.

mbramson replied:
It's kind of sort of used in the GroupPairController. There's actually a bit of refactoring that should occur there as well to mirror the fact that technically in the PairController, it's now legal to be in multiple pairs in one week for one group. This results in some buggy behavior in the GroupPairController, which is fine for now since this state can only be reached if something has gone horribly wrong, or someone is poking the database manually.

mbramson commented on web/controllers/pair_controller.ex:
```diff
@@ -29,30 +29,37 @@ defmodule Pairmotron.PairController do
+            pairs_and_retros = pairs
+              |> pairs_containing_user(user)
+              |> add_retros_to_pairs_for_user(user)
+            new_groups_with_pairs = groups_with_pairs ++ [{group, pairs_and_retros}]
```
There's definitely duplication between these two cases. I should probably extract that.

ericworkman replied:
Yeah, these are identical except for the message handling, right? Would it be better to adjust `fetch_or_gen` to be consistent about how many items it returns in the :ok and :error tuples?

mbramson replied:
That would make sense actually. In the :ok case, it could just return an empty list of messages, that way duplication goes away without any extraction.

ericworkman replied:
Yeah, I think that is reasonable.

mbramson reviewed (commented):
Thoughts and commentary

mbramson committed:
Make PairMaker.fetch_or_gen return type consistent

Return {:ok, pairs, nil} instead of {:ok, pairs} to make processing the
output easier, since the error output is {:error, pairs, message}

mbramson committed:
Move assign_current_user_pair_retro_for_week

This function used to be called in two unique places, but is now only
called in the GroupPairController, so it was moved there. It will
probably be replaced entirely in a subsequent commit anyhow.

mbramson committed:
Add private pair_session type for clarity

mbramson committed:
Refactor function for clarity

mbramson commented:
@ericworkman, I believe that I've addressed your concerns. Take a look!

ericworkman reviewed (approved):
Looks great! Awesome job!
